packages/components/src/navigator/navigator-provider/component.tsx

```diff
@@ -0,0 +1,96 @@
+/**
+ * External dependencies
+ */
+// eslint-disable-next-line no-restricted-imports
+import type { Ref } from 'react';
+
+/**
+ * WordPress dependencies
+ */
+import { useState } from '@wordpress/element';
+
+/**
+ * Internal dependencies
+ */
+import {
+	contextConnect,
+	useContextSystem,
+	WordPressComponentProps,
+} from '../../ui/context';
+import { View } from '../../view';
+import { NavigatorContext } from '../context';
+import type { NavigatorProviderProps, NavigatorPath } from '../types';
+
+function NavigatorProvider(
+	props: WordPressComponentProps< NavigatorProviderProps, 'div' >,
+	forwardedRef: Ref< any >
+) {
+	const { initialPath, children, ...otherProps } = useContextSystem(
+		props,
+		'NavigatorProvider'
+	);
+
+	const [ path, setPath ] = useState< NavigatorPath >( {
+		path: initialPath,
+	} );
+
+	return (
+		<View ref={ forwardedRef } { ...otherProps }>
```

Comment:
Why do we show a view at all, this is probably going to create an extra div for nothing.

Reply:
Having a wrapping `div` around all the "screens" made sense to me, and I think is going to be necessary to fix the scrollbar bug (flagged initially [here](https://github.com/WordPress/gutenberg/pull/34904#pullrequestreview-760854463)) by setting `overflow: hidden` on it.

On top of that, it offers a component to attach a `forwardedRef` to, together with all of the remaining props that may be passed to this component (given that we mark its props as `WordPressComponentProps`).

Reply:
I was thinking of applying `overflow: hidden` only while the transition is in progress (I have already a working proof of concept on my machine) — which shouldn't cause issues with modals & popovers.

I also spent some time looking into alternative solutions, but so far I haven't found anything better.

Reply:
My argument for a wrapper would stem from wanting to [animate the exits](35311) of `navigatorScreen`s and that can't work reliably without one.

I don't think Marco is proposing we add `overflow: hidden` was going to be added in this PR so we don't need to debate that here. I think it's going to be the solution but maybe not applied in the component itself or if so with a prop to control it.

Reply:
> I don't think Marco is proposing we add overflow: hidden in this PR

That is true (we should be tacking that in a separate PR), but at the same time, it is also one of the reasons for introducing a wrapper `div`.

By having `NavigatorProvider` rendering a wrapper `div`, we would be in control of a root DOM element and its layout more explicitly (we could control its overflow, but also is display settings, its CSS [containment](https://developer.mozilla.org/en-US/docs/Web/CSS/contain), ...)

Reply:
> (I have already a working proof of concept on my machine) 

I've refined my working proof and opened #35317 (which is based on top of this PR)